appserver/orb/orb-iiop/src/main/java/org/glassfish/enterprise/iiop/impl/IIOPSSLSocketFactory.java

```diff
@@ -651,34 +618,82 @@ private Socket createSSLSocket(String host, int port)
             _logger.log(Level.FINE, "Merged socket ciphers: " + cList);
         }
 
-        return (String[])cList.toArray(new String[cList.size()]);
+        return (String[]) cList.toArray(new String[cList.size()]);
     }
 
     /**
      * Check whether given cipherInfo belongs to given protocol.
+     *
      * @param cipherInfo
      * @param ssl2Enabled
      * @param ssl3Enabled
      * @param tlsEnabled
      */
-    private boolean isValidProtocolCipher(CipherInfo cipherInfo,
-            boolean ssl2Enabled, boolean ssl3Enabled, boolean tlsEnabled, boolean tlsEnabled11, boolean tlsEnabled12) {
+    private boolean isValidProtocolCipher(CipherInfo cipherInfo, boolean ssl2Enabled, boolean ssl3Enabled,
+            boolean tlsEnabled, boolean tlsEnabled11, boolean tlsEnabled12) {
         return (tlsEnabled && cipherInfo.isTLS() ||
                 tlsEnabled11 && cipherInfo.isTLS() ||
                 tlsEnabled12 && cipherInfo.isTLS() ||
                 ssl3Enabled && cipherInfo.isSSL3() ||
                 ssl2Enabled && cipherInfo.isSSL2());
     }
 
+    /**
+     * Checks whether SO_KEEPALIVE should be enabled on a Socket and enables it if it should be. Checks for the
+     * presence of a property on the IIOP listener and globally, preferring the value set in the listener.
+     *
+     * @param socket The socket to potentially enable SO_KEEPALIVE on
+     * @throws SocketException If there was an error enabling SO_KEEPALIVE on the socket.
+     */
+    private void enableSOKeepAliveAsRequired(Socket socket) throws SocketException {
+        boolean shouldSet = false;
+
+        // For each listener, find one with a matching port
+        for (IiopListener iiopListener : IIOPUtils.getInstance().getIiopService().getIiopListener()) {
+            if (Integer.valueOf(iiopListener.getPort()) == socket.getPort()) {
+                // Check for the property globally before checking on the specific listener
+                if (Boolean.getBoolean(SO_KEEPALIVE)) {
+                    // If it's present globally, check if we've actually set the property on the listener
+                    if (iiopListener.getPropertyValue(SO_KEEPALIVE) != null) {
+                        // If it has been set, check if we should override
+                        if (Boolean.valueOf(iiopListener.getPropertyValue(SO_KEEPALIVE))) {
+                            shouldSet = true;
+                        }
+                    } else {
+                        // If the property wasn't set on the listener, go with the global setting
+                        shouldSet = true;
+                    }
+                    break;
+                } else {
+                    // If it wasn't set globally, just check if it's set on the listener
+                    if (iiopListener.getPropertyValue(SO_KEEPALIVE) != null
```

Comment:
same here, null check isn't necessary

Reply:
The null check is for checking if it's actually set vs. checking if it's false. I *want* it to tell me if it's null, rather than telling me if it's false - I'm using the null check essentially as a presence check.

The reasoning for this is that value can be set in two places, globally via system property (`Boolean.getBoolean`) and specifically on the listener itself (`iiopListener.getPropertyValue`). So I'm checking what the value is globally, and then overriding that *if* it's set on the specific listener - if I just used the Boolean value it would override the global value with a "false" even if it hadn't been set.